crates/generate/src/funcs.rs

```diff
@@ -209,102 +204,25 @@ fn marshal_arg(
                 let len_name = names.func_len_binding(&param.name);
                 let name = names.func_param(&param.name);
                 quote! {
-                    let num_elems = match memory.ptr::<u32>(#len_name as u32) {
-                        Ok(p) => match p.as_ref() {
-                            Ok(r) => r,
-                            Err(e) => {
-                                #error_handling
-                            }
-                        }
-                        Err(e) => {
-                            #error_handling
-                        }
-                    };
-                    let #name: wiggle_runtime::GuestString<#lifetime> = match memory.ptr::<u8>(#ptr_name as u32) {
-                        Ok(p) => match p.array(*num_elems) {
-                            Ok(s) => s.into(),
-                            Err(e) => {
-                                #error_handling
-                            }
-                        }
-                        Err(e) => {
-                            #error_handling
-                        }
-                    };
+                    let #name = wiggle_runtime::GuestPtr::<#lifetime, str>::new(memory, (#ptr_name as u32, #len_name as u32));
                 }
             }
         },
-        witx::Type::Pointer(pointee) => {
+        witx::Type::Pointer(pointee) | witx::Type::ConstPointer(pointee) => {
```

Comment:
This is cool since as you've pointed out we've got less surface to cover since now we operate with only one pointer type. But at the same time, do we want to splice the meaning of `(@witx const_pointer _)` and `(@witx pointer _)` into one entity? Personally, I'm OK with either, but the original design proposed by @pchickey reminded me of the distinction made in Rust where we've got `*const _` and `*mut _` (which is not say you cannot easily convert one into another!).

Reply:
Ah sorry yeah I didn't call this out well in the PR description, and I left a [late comment about this too](https://github.com/kubkon/wiggle/pull/36#issuecomment-594754160). Basically I don't think that mutability here buys us anything in terms of actual memory safety. For that reason I wanted to start from the minimal "known safe" API surface area and expand outward from there. 

Now that being said there's certainly more kinds of safety than memory safety and having the mutability distinction could definitely help catch other bugs. All that's needed is to increase the API surface area and provide nice conversions from `GuestPtr` to `GuestPtrMut` (like a `*const` to `*mut` cast).

Certainly don't mind adding it if it's still desired.

Reply:
As I said, I'm fine with just one pointer type, especially for now given that we're still in our infancy with this crate (I think). But I'll leave the final call with @pchickey.